File: app/src/ultimate/java/com/clockmods/widget/store/WidgetConfigStore.java
package com.clockmods.widget.store;
import android.content.Context;
import android.content.SharedPreferences;
import com.clockmods.widget.model.*;
import org.json.JSONObject;
import java.util.ArrayList;
import java.util.List;
/** JSON codec is also usable without Android, for migrations and external preset tooling. */
public final class WidgetConfigStore {
    private final SharedPreferences prefs;
    public WidgetConfigStore(Context context) { this(context.getSharedPreferences("clockmods_widgets", Context.MODE_PRIVATE)); }
    public WidgetConfigStore(SharedPreferences prefs) { this.prefs = prefs; }
    public WidgetConfig getOrDefault(int id, WidgetKind kind) {
        String raw = prefs.getString("widget_" + id, null);
        WidgetConfig c = decode(raw, id, kind);
        if (raw != null && !encode(c).equals(raw)) save(c);
        return c;
    }
    public void save(WidgetConfig c) { prefs.edit().putString("widget_" + c.appWidgetId, encode(c)).apply(); }
    public void delete(int id) { prefs.edit().remove("widget_" + id).apply(); }
    public boolean contains(int id) { return prefs.contains("widget_" + id); }
    public List<WidgetConfig> getAll() {
        List<WidgetConfig> out = new ArrayList<>();
        for (String key : prefs.getAll().keySet()) {
            if (!key.startsWith("widget_")) continue;
            try {
                int id = Integer.parseInt(key.substring(7));
                String raw = prefs.getString(key, null);
                WidgetKind kind = WidgetKind.DIGITAL;
                try { kind = WidgetKind.valueOf(new JSONObject(raw).optString("kind")); } catch (Exception ignored) { }
                out.add(decode(raw, id, kind));
            } catch (NumberFormatException ignored) { }
        }
        return out;
    }
    public List<Integer> getIdsUsingWeather() {
        List<Integer> out = new ArrayList<>();
        for (WidgetConfig c : getAll()) if (c.kind == WidgetKind.WEATHER) out.add(c.appWidgetId);
        return out;
    }
    public static String encode(WidgetConfig c) {
        try {
            JSONObject j = new JSONObject().put("schemaVersion", c.schemaVersion);
            j.put("appWidgetId", c.appWidgetId);
            j.put("kind", c.kind.name());
            j.put("themeId", c.themeId);
            j.put("timeZoneId", c.timeZoneId);
            j.put("useSystemTimeZone", c.useSystemTimeZone);
            j.put("useSystemTimeFormat", c.useSystemTimeFormat);
            j.put("use24Hour", c.use24Hour);
            j.put("showSeconds", c.showSeconds);
            j.put("showDate", c.showDate);
            j.put("showWeekday", c.showWeekday);
            j.put("showLunar", c.showLunar);
            j.put("showWeatherDescription", c.showWeatherDescription);
            j.put("showLocation", c.showLocation);
            j.put("backgroundAlpha", c.backgroundAlpha);
            j.put("textScale", c.textScale);
            j.put("tapAction", c.tapAction);
            j.put("updatedAt", c.updatedAt);
            j.put("darkText", c.darkText);
            j.put("fontId", c.fontId);
            return j.toString();
        } catch (org.json.JSONException e) { throw new IllegalArgumentException(e); }
    }
    public static WidgetConfig decode(String raw, int id, WidgetKind kind) {
        WidgetConfig.Builder b = WidgetConfig.builder(id, kind);
        WidgetConfig d = b.build();
        if (raw == null) return d;
        try {
            JSONObject j = new JSONObject(raw);
            j = migrate(j, j.optInt("schemaVersion", 0));
            b.themeId(j.optString("themeId", d.themeId));
            b.timeZoneId(j.optString("timeZoneId", d.timeZoneId));
            b.useSystemTimeZone(j.optBoolean("useSystemTimeZone", d.useSystemTimeZone));
            b.useSystemTimeFormat(j.optBoolean("useSystemTimeFormat", d.useSystemTimeFormat));
            b.use24Hour(j.optBoolean("use24Hour", d.use24Hour));
            b.showSeconds(j.optBoolean("showSeconds", d.showSeconds));
            b.showDate(j.optBoolean("showDate", d.showDate));
            b.showWeekday(j.optBoolean("showWeekday", d.showWeekday));
            b.showLunar(j.optBoolean("showLunar", d.showLunar));
            b.showWeatherDescription(j.optBoolean("showWeatherDescription", d.showWeatherDescription));
            b.showLocation(j.optBoolean("showLocation", d.showLocation));
            b.backgroundAlpha(j.optInt("backgroundAlpha", d.backgroundAlpha));
            b.textScale((float) j.optDouble("textScale", d.textScale));
            b.tapAction(j.optString("tapAction", d.tapAction));
            b.updatedAt(j.optLong("updatedAt", d.updatedAt));
            b.darkText(j.optBoolean("darkText", d.darkText));
            b.fontId(j.optString("fontId", d.fontId));
            return b.build();
        } catch (Exception ignored) { return d; }
    }
    private static JSONObject migrate(JSONObject source, int fromVersion) throws org.json.JSONException {
        if (fromVersion > 1) throw new IllegalArgumentException("Unsupported widget schema");
        if (fromVersion == 0) source.put("schemaVersion", 1);
        return source;
    }
}

File: app/src/ultimate/java/com/clockmods/widget/store/WidgetConfigStore.kt
package com.clockmods.widget.store

import android.content.Context
import android.content.SharedPreferences
import com.clockmods.widget.model.WidgetConfig
import com.clockmods.widget.model.WidgetKind
import org.json.JSONObject

/** JSON codec is also usable without Android, for migrations and external preset tooling. */
class WidgetConfigStore {
    private val prefs: SharedPreferences

    constructor(context: Context) : this(
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE),
    )

    constructor(prefs: SharedPreferences) {
        this.prefs = prefs
    }

    fun getOrDefault(id: Int, kind: WidgetKind): WidgetConfig {
        val raw = prefs.getString(key(id), null)
        val config = decode(raw, id, kind)
        if (raw != null && encode(config) != raw) save(config)
        return config
    }

    fun save(config: WidgetConfig) {
        prefs.edit().putString(key(config.appWidgetId), encode(config)).apply()
    }

    fun delete(id: Int) {
        prefs.edit().remove(key(id)).apply()
    }

    fun contains(id: Int): Boolean = prefs.contains(key(id))

    fun getAll(): List<WidgetConfig> {
        val out = ArrayList<WidgetConfig>()
        for (entryKey in prefs.all.keys) {
            if (!entryKey.startsWith(KEY_PREFIX)) continue
            try {
                val id = entryKey.substring(KEY_PREFIX.length).toInt()
                val raw = prefs.getString(entryKey, null)
                val kind = try {
                    WidgetKind.valueOf(JSONObject(raw ?: "").optString("kind"))
                } catch (_: Exception) {
                    WidgetKind.DIGITAL
                }
                out += decode(raw, id, kind)
            } catch (_: NumberFormatException) {
                // Ignore unrelated or damaged preference keys, matching the legacy behavior.
            }
        }
        return out
    }

    fun getIdsUsingWeather(): List<Int> = getAll()
        .filter { it.kind == WidgetKind.WEATHER }
        .map { it.appWidgetId }

    companion object {
        private const val PREFS_NAME = "clockmods_widgets"
        private const val KEY_PREFIX = "widget_"

        private fun key(id: Int): String = KEY_PREFIX + id

        @JvmStatic fun encode(config: WidgetConfig): String = JSONObject()
            .put("schemaVersion", config.schemaVersion)
            .put("appWidgetId", config.appWidgetId)
            .put("kind", config.kind.name)
            .put("themeId", config.themeId)
            .put("timeZoneId", config.timeZoneId)
            .put("useSystemTimeZone", config.useSystemTimeZone)
            .put("useSystemTimeFormat", config.useSystemTimeFormat)
            .put("use24Hour", config.use24Hour)
            .put("showSeconds", config.showSeconds)
            .put("showDate", config.showDate)
            .put("showWeekday", config.showWeekday)
            .put("showLunar", config.showLunar)
            .put("showWeatherDescription", config.showWeatherDescription)
            .put("showLocation", config.showLocation)
            .put("backgroundAlpha", config.backgroundAlpha)
            .put("textScale", config.textScale)
            .put("tapAction", config.tapAction)
            .put("updatedAt", config.updatedAt)
            .put("darkText", config.darkText)
            .put("fontId", config.fontId)
            .toString()

        @JvmStatic fun decode(raw: String?, id: Int, kind: WidgetKind): WidgetConfig {
            val builder = WidgetConfig.builder(id, kind)
            val defaults = builder.build()
            if (raw == null) return defaults
            return try {
                val source = JSONObject(raw)
                val json = migrate(source, source.optInt("schemaVersion", 0))
                builder
                    .themeId(json.optString("themeId", defaults.themeId))
                    .timeZoneId(json.optString("timeZoneId", defaults.timeZoneId))
                    .useSystemTimeZone(json.optBoolean("useSystemTimeZone", defaults.useSystemTimeZone))
                    .useSystemTimeFormat(json.optBoolean("useSystemTimeFormat", defaults.useSystemTimeFormat))
                    .use24Hour(json.optBoolean("use24Hour", defaults.use24Hour))
                    .showSeconds(json.optBoolean("showSeconds", defaults.showSeconds))
                    .showDate(json.optBoolean("showDate", defaults.showDate))
                    .showWeekday(json.optBoolean("showWeekday", defaults.showWeekday))
                    .showLunar(json.optBoolean("showLunar", defaults.showLunar))
                    .showWeatherDescription(json.optBoolean("showWeatherDescription", defaults.showWeatherDescription))
                    .showLocation(json.optBoolean("showLocation", defaults.showLocation))
                    .backgroundAlpha(json.optInt("backgroundAlpha", defaults.backgroundAlpha))
                    .textScale(json.optDouble("textScale", defaults.textScale.toDouble()).toFloat())
                    .tapAction(json.optString("tapAction", defaults.tapAction))
                    .updatedAt(json.optLong("updatedAt", defaults.updatedAt))
                    .darkText(json.optBoolean("darkText", defaults.darkText))
                    .fontId(json.optString("fontId", defaults.fontId))
                    .build()
            } catch (_: Exception) {
                defaults
            }
        }

        private fun migrate(source: JSONObject, fromVersion: Int): JSONObject {
            require(fromVersion <= 1) { "Unsupported widget schema" }
            if (fromVersion == 0) source.put("schemaVersion", 1)
            return source
        }
    }
}
